pkg/options/namespace_list/parser.go

```diff
@@ -0,0 +1,178 @@
+/*
+Copyright 2021 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package namespacelist
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// Parser parses the syntax for namespace lists
+type Parser struct {
+	source string
+	pos    int
+}
+
+// UnexpectedSyntaxError is an error which is returned when the value being parsed does not match the syntax
+type UnexpectedSyntaxError struct {
+	char string
+	pos  int
+}
+
+// Error returns the error as a human-readable string
+func (err UnexpectedSyntaxError) Error() string {
+	return fmt.Sprintf("syntax error, unexpected %v at col %v", err.char, err.pos)
+}
+
+// Parse parses the source into a map
```

Comment:
Can we document what the resulting map contains or how it looks like?

Alternatively, we can return a struct instead of a map. That should also make it easier to understand the contents by just reading the code.

Reply:
Would the struct be something like below? I much rather use the struct approach than document it. Code should explain itself as much as possible.

```go
type NamespaceList []struct {
    Namespace string
    Labels []string
}
```

Reply:
Or maybe something like 

```go
type NamespaceMatcher struct {
    Name string
    Labels []string
}
```

Reply:
I believe using "Matcher" in this case insinuates that the struct is meant to do more than be a data structure.

Reply:
Now that you brought this up, I don't think that would be a bad idea. We can filter namespaces by labels by passing the labels to the api server, but we can't really filter by prefix when using wildcards. So I assume that logic will have to be in KSM and the NamespaceMatcher can have a `Matches(string) bool` or `Matches(corev1.Namespace) bool` function to facilitate that. It's done similarly here, albeit through an interface: https://pkg.go.dev/k8s.io/apimachinery/pkg/labels#Selector

Do you have some idea of how you intend to do the filtering?